test/select.jl

```diff
@@ -0,0 +1,838 @@
+    @test names(df2) == [:x2, :x1, :x3, :x4, :r1, :r2]
+    @test df.x2 == df2.x2
+    @test df.x2 !== df2.x2
+    @test df.x1 == df2.x4
+    @test df.x4 !== df2.x1
+    @test df2.r1 == df.x1 .^ 2
+    @test df2.r1 == df2.r2
+    @test df2.x1 == df.x1 + df.x2
+    @test df2.x3 == df.x1 ./ df.x2
+
+    @test select(df, [:x1, :x1] => +) == DataFrame(Symbol("x1_x1_+") => 2*df.x1)
+    @test select(df, [1, 1] => +) == DataFrame(Symbol("x1_x1_+") => 2*df.x1)
+
+    df2 = select(df, :x2, :, :x1 => ByRow(x -> x^2) => :r1, :x1 => (x -> x .^ 2) => :r2,
+                 [:x1, :x2] => (+) => :x1, 1:2 => ByRow(/) => :x3, :x1 => :x4, copycols=false)
+    @test names(df2) == [:x2, :x1, :x3, :x4, :r1, :r2]
+    @test df.x2 === df2.x2
+    @test df.x1 === df2.x4
+    @test df2.r1 == df.x1 .^ 2
+    @test df2.r1 == df2.r2
+    @test df2.x1 == df.x1 + df.x2
+    @test df2.x3 == df.x1 ./ df.x2
+
+    x1, x2, x3, x4 = df.x1, df.x2, df.x3, df.x4
+    select!(df, :x2, :, :x1 => ByRow(x -> x^2) => :r1, :x1 => (x -> x .^ 2) => :r2,
+            [:x1, :x2] => (+) => :x1, 1:2 => ByRow(/) => :x3, :x1 => :x4)
+    @test names(df2) == [:x2, :x1, :x3, :x4, :r1, :r2]
+    @test x2 === df.x2
+    @test x1 === df.x4
+    @test df.r1 == x1 .^ 2
+    @test df.r1 == df.r2
+    @test df.x1 == x1 + x2
+    @test df.x3 == x1 ./ x2
+end
+
+@testset "select and select! reserved return values" begin
+    df = DataFrame(x=1)
+    df2 = copy(df)
+    for retval in [df2, (a=1, b=2), df2[1, :], ones(2,2)]
+        @test_throws ArgumentError select(df, :x => x -> retval)
+        @test_throws ArgumentError select(df, :x => x -> retval, copycols=false)
+        @test_throws ArgumentError select!(df, :x => x -> retval)
+        @test select(df, :x => ByRow(x -> retval)) == DataFrame(x_function = [retval])
+        cdf = copy(df)
+        select!(cdf, :x => ByRow(x -> retval))
+        @test cdf == DataFrame(x_function = [retval])
+    end
+
+    for retval in [(1, 2), ones(2,2,2)]
+        @test select(df, :x => x -> retval) == DataFrame(x_function = [retval])
+        @test select(df, :x => ByRow(x -> retval)) == DataFrame(x_function = [retval])
+        cdf = copy(df)
+        select!(cdf, :x => x -> retval)
+        @test cdf == DataFrame(x_function = [retval])
+        cdf = copy(df)
+        select!(cdf, :x => ByRow(x -> retval))
+        @test cdf == DataFrame(x_function = [retval])
+    end
+end
+
+@testset "select and select! empty selection" begin
+    df = DataFrame(rand(10, 4))
+    x = [1,2,3]
+
+    @test select(df, r"z") == DataFrame()
+    @test select(df, r"z" => () -> x) == DataFrame(_function=x)
+    @test select(df, r"z" => () -> x)[!, 1] === x # no copy even for copycols=true
+    @test_throws MethodError select(df, r"z" => x -> 1)
+    @test_throws ArgumentError select(df, r"z" => ByRow(rand))
+
+    @test select(df, r"z", copycols=false) == DataFrame()
+    @test select(df, r"z" => () -> x, copycols=false) == DataFrame(_function=x)
```

Comment:
This looks a bit weird. Would it make sense to drop the `_`? IOW the rule would be that each variable name is followed with `_`.

Reply:
OK.